Signal/src/ViewControllers/LoadingViewController.swift

```diff
@@ -0,0 +1,102 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import PromiseKit
+
+// The initial presentation is intended to be indistinguishable from the Launch Screen.
+// After a delay we present some "loading" UI so the user doesn't think the app is frozen.
+@objc
+public class LoadingViewController: UIViewController {
+
+    var logoView: UIImageView!
+    var topLabel: UILabel!
+    var bottomLabel: UILabel!
+
+    override public func loadView() {
+        self.view = UIView()
+        view.backgroundColor = UIColor.ows_materialBlue
+
+        self.logoView = UIImageView(image: #imageLiteral(resourceName: "logoSignal"))
+        view.addSubview(logoView)
+
+        logoView.autoCenterInSuperview()
+        logoView.autoPinToSquareAspectRatio()
+        logoView.autoMatch(.width, to: .width, of: view, withMultiplier: 1/3)
+
+        self.topLabel = buildLabel()
+        topLabel.alpha = 0
+        topLabel.font = UIFont.ows_dynamicTypeTitle2
+        topLabel.text = NSLocalizedString("DATABASE_VIEW_OVERLAY_TITLE", comment: "Title shown while the app is updating its database.")
+
+        self.bottomLabel = buildLabel()
+        bottomLabel.alpha = 0
+        bottomLabel.font = UIFont.ows_dynamicTypeBody
+        bottomLabel.text = NSLocalizedString("DATABASE_VIEW_OVERLAY_SUBTITLE", comment: "Subtitle shown while the app is updating its database.")
+
+        let labelStack = UIStackView(arrangedSubviews: [topLabel, bottomLabel])
+        labelStack.axis = .vertical
+        labelStack.alignment = .center
+        labelStack.spacing = 8
+        view.addSubview(labelStack)
+
+        labelStack.autoPinEdge(.top, to: .bottom, of: logoView, withOffset: 20)
+        labelStack.autoPinLeadingToSuperviewMargin()
+        labelStack.autoPinTrailingToSuperviewMargin()
+        labelStack.setCompressionResistanceHigh()
+        labelStack.setContentHuggingHigh()
+    }
+
+    var isShowingTopLabel = false
+    var isShowingBottomLabel = false
+    override public func viewDidAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        // We only show the "loading" UI if it's a slow launch. Otherwise this ViewController
+        // shoudl be indistinguishable from the launch screen.
+        let kTopLabelThreshold: TimeInterval = 5
+        DispatchQueue.main.asyncAfter(deadline: .now() + kTopLabelThreshold) { [weak self] in
```

Comment:
* 5 seconds seems long before showing this warning. I'd suggest 2?
* The reason I didn't use a timer before is because the main thread might be blocked on a db read/write or whatever.  It shouldn't be, but this seems like an unsafe assumption.

Reply:
> 5 seconds seems long before showing this warning. I'd suggest 2?

I'm not confident that our normal startup is under 2 seconds for most devices. We definitely don't want to risk showing this every time someone launches. In fact, I actually had the opposite concern, that 5 seconds could be too soon for old slow devices.

> The reason I didn't use a timer before is because the main thread might be blocked on a db read/write or whatever. It shouldn't be, but this seems like an unsafe assumption.

I suppose that's true for any code that uses the main thread anywhere. Do you have a proposal?